lib/strmalloc.man
.\" RCSID @(#)$Id: strmalloc.man,v 1.1 2002/07/18 22:23:57 rkowen Exp $
.TH "STRMALLOC" "3rko" "01 Sept 1999"
.SH NAME
strmalloc \- adds a strdup-like routine for copying strings
strfree \- frees the string memory storage
.SH SYNOPSIS

 #include "strmallloc.h"

 /* old interface */
 char	*strnmalloc(	char const *in, size_t n);
 char	*strmalloc(	char const *in);
 int	 strfree(	char **str);
 
 /* new interface */
 char	*str_nmalloc(	char const *in, size_t n);
 char	*str_malloc(	char const *in);
 int	 str_free(	char **str);
 size_t	 str_sizeof(	char const *str);
 char	*str_ncpy(	char **str, char *in, size_t n);
 char	*str_cpy(	char **str, char *in);
 char	*str_ncat(	char **str, char *in, size_t n);
 char	*str_cat(	char **str, char *in);

.SH DESCRIPTION
.I str_malloc
adds a strdup-like function for routinely allocating memory
and copying a string to some other location.
The only difference is that
.I str_malloc
allocates memory on the heap via
.I malloc()
instead of the stack.  Hence the memory needs to be freed
explicitly.  Use the associated
.I str_free
to free up the memory.   First it will wipe out the memory area
and then call
.IR free() .
.P
.TP 15
.I str_malloc
copy a string and allocate memory for it, along with some internal
data.
.TP
.I strmalloc
same as above.
.TP
.I str_nmalloc
same as
.I str_malloc
but the amount to allocate and copy is specified.  Can be over or under
the string length.
.TP
.I strnmalloc
same as above.
.TP
.I str_free
free a string object, clearing memory as much as possible.
.TP
.I strfree
same as above.
.TP
.I str_sizeof
gives the allocated length of the string.
.TP
.I str_cpy
like strcpy, but for string objects.
Will automatically extend the allocated space if needed.
.TP
.I str_ncpy
like strncpy, but for string objects, and can be over
or under the string length.
Will automatically extend the allocated space if needed.
.TP
.I str_cat
like strcat, but for string objects.
Will automatically extend the allocated space if needed.
.TP
.I str_ncat
like strncat, but for string objects, and can be over
or under the string length.
Will automatically extend the allocated space if needed.

.SH NOTES
The allocated memory will actually be a few bytes larger, since
.I str_malloc
prepends the length of the string and a identifiable tag string.

.SH DIAGNOSTICS
If any of the routines that return (char *) return a NULL,
then some sort of error occured.
.P
.I str_sizeof
returns (size_t) -1 if an invalid string object, note that (size_t)
may be an unsigned integer type and that the return value may not
be negative, but a rather huge positive value.  Always test for
(size_t) -1 by direct comparison.
.P
.I str_free()
only returns an error if the user sent an invalid string object
(i.e. user error).
Details can be queried from
.I rkoerrno
and
.I rkostrerror()
if RKOERROR is enabled.
However, it's not really necessary for the same reasons that
.I free()
doesn't return a status.

.SH SEE ALSO
string(3), strdup(3), malloc(3), free(3), rkoerror(3rko)

.SH AUTHOR
R.K.Owen,Ph.D.

.KEY WORDS


NEWS
* Note: Implied Consent -
  If you send in a code patch or fix of one type or another you consent
  to have your name and email address listed here unless you explicitly
  note otherwise.  Whether your name & email is actually listed depends
  on how extensive the change.  (If it's a minor code, document change,
  or spelling correction your name will probably not be listed.)

****************************   Release 3.2 ******************************
Modules 3.2.9
	* Fix the coexistence of load/remove flags (Martin Siegert)
	* Fixed an overlapping string copy (Wilson Snyder & Orion Poplawski)
	* Isolated calls to the new interpreter and get & set env.vars.
	* Optionally use Tcl memory check routines for malloc/realloc.
	* Fixed the "module purge" memory corruption (Poor Yorick)
	* Added support for Ruby (Tammo Tjarks)

Modules 3.2.8
Oct  1, 2010				R.K. Owen (rk@owen.sj.ca.us)
	* Handle the .modulerc/.version in deep modulefile dirs,
	  Patches thanks to Lindsay Todd, and others.
	* Note that .version takes precedence over .modulerc files.
	* Renamed init/python to init/python.py and updated with
	  anonymous patch to eliminate 'sed' references
	* Renamed init/perl to init/perl.pm
	* Added brief section to module.1 on invoking modules for perl & python
	* Added Nathaniel Waisbrot's patch for CMake support with tests.
	* Fixed the MANPATH in modulefiles/modules to use configure generated
	  values.
	* Added the configure --with-man-path so a system specific MANPATH
	  could be used.  Some "man"s radically change their search path
	  when a MANPATH exists and when not.
	  The default is "/usr/man:/usr/share/man" (when both exists).
	* Purging a purged module list does not result in an error.
	* Fixed the configure script for searching for tclConfig.sh
	  and handle --without-x in the tests.

Modules 3.2.7
Jul 30, 2009				R.K. Owen (rk@owen.sj.ca.us)
	* Have init/bash export the module function (thanks to Roy Dragseth)
	* Fixed swap bug when prepending or appending multiple times
	  (fix thanks to Gavin Walker), added tests for same.
	* Have purge remove loaded modules in reverse order
	  (patch contributed by carriees)
	* Added a batch of patches from Gerrit Renker
	  - Goes through each of the two manpages, bringing them up to date.
	  - Fix an error in configure.ac caused by misspelling
	  - Fix typos in help messages and source-code comments
	  - Fix the setup of $MODULESPATH to not have a trailing ':'
	  - Fix segmentation violation occurring when the shelltype is absent
	  - Fix a condition where 'make uninstall' does not cleanse a directory
	  - Fix a build problem which set the executable bit on 'data' files
	  - Cleans up noise in '.modulespath' file
	  - Replaces external sed filter script with internal Makefile.am rule
	  - Adds missing 'zsh' case in etc/global
	  - simplify (t)csh initialisation
	  - Add bash auto-completion for the 'module' command
	  - Adds new 'chdir' modulefile command (sets cwd on module load)
	* versions a & b fixed some bugs

Modules 3.2.6
Oct 29, 2007				R.K. Owen (rk@owen.sj.ca.us)
	* Fixed the locate_module.c:realloc() calls which didn't properly
	  take the element size into account (Fix thanks to Ron Isaacson).
Modules 3.2.5
Feb 14, 2007				R.K. Owen (rk@owen.sj.ca.us)
	* Wrapped all malloc() calls to make sure the size is never 0.
	  This fixes a problem spotted by Francesco Nigro under AIX,
	  and diagnosed by G Lehnert.
	* Updated tests when domainname is unknown.
Modules 3.2.4
Feb  2, 2007				R.K. Owen (rk@owen.sj.ca.us)
	* Fixed bug where "empty" directory paths in the MODULEPATH env.var.
	  needed to be handled gracefully.
Modules 3.2.3
Jun  1, 2006				R.K. Owen (rk@owen.sj.ca.us)
	* Allow version names to be substrings of other version names
	  thanks to dgoldendias at users.sourceforge.net
	* Fixed compiling bug when --enable-log was specified.
	* Made some minor changes to the syslogging and tests so that
	  they all pass.
	* the set-alias command will for Bourne shell derivatives
	  substitute the safe ${1+"$@"} for $*.
	* Added the --delim option to prepend-path and append-path to
	  allow arbitrary delimiters instead of the colon ':'
	* Fixed build problem with regards to the init/* files
	  (solution due to Thomas Zeiser)
Modules 3.2.2
May  4, 2006				R.K. Owen (rk@owen.sj.ca.us)
	* Replaced all strtok() calls with an internal version.
	* Now can append/prepend and empty string '' or colons ':'
	* Added more --with-tclx-* options in case tclxConfig.sh
	  can not be found, but the TclX lib does exist.
	* Expanded the search path for tclConfig.sh.
	* Updated the TclX init call for 8.4
Modules 3.2.1
Feb 13, 2006				R.K. Owen (rk@owen.sj.ca.us)
	* Catch 'exit N' within modulefiles, use this as an exit return code.
	  Has the behavior of loading the modulefile up to the point of exit.
	* For bash, a non-zero exit will append a test that returns false.
	* Added tests for "break" and "exit N"
	* various other bug fixes
Modules 3.2.0
Jan 19, 2006				R.K. Owen (rk@owen.sj.ca.us)
	* default for cache is now "no"
	* Added version_test() to testsuite to make some of the
	  tests conditional on the values generated by --version.
	* Reworked every configure-based file
	  - the configure options help is formatted with AC_HELP_STRING
	  - uses automake to generate most of the Makefile.in
	  - renamed configure.in to configure.ac per autoconf
	  - Versioning is optional now with a configure option
	  - reworked the init scripts for versioning or not
	  - eliminated acconfig.h and put templates in configure.ac
	  - make install now honors the DESTDIR variable for package builds
	  - honors bindir for modulecmd location
	   (but the init scripts are only correct if no versioning)
	* Collected the usage info and now it can be displayed with
	  the --help or -H option (same as help)
	* --version now gives most of the optional configure parameters
	  for easy debugging of problems
	* Made the output deterministic (sorted) and reworked tests also

lib/strmalloc.c
static const char RCSID[]="@(#)$Id: strmalloc.c,v 1.1 2002/07/18 22:23:57 rkowen Exp $";
static const char AUTHOR[]="@(#)strmalloc 1.0 09/01/1999 R.K.Owen,Ph.D.";
/* str_malloc	- adds a strdup-like routine for copying strings.
 * str_nmalloc	- same as above with over allocation.
 * str_free	- frees the string memory storage.
 * str_sizeof	- returns the maximum string size.
 * str_cpy	- safely copies the input string to the str_malloc object
 * str_cat	- safely cats the input string onto the str_malloc object
 *
 * by R.K. Owen,Ph.D.   1999/09/01
 */
/** ** Copyright *********************************************************** **
 ** 									     **
 ** Copyright 1999 by R.K.Owen,Ph.D.		                      	     **
 ** last known email: librko@kooz.sj.ca.us				     **
 **                   rk@owen.sj.ca.us					     **
 ** see LICENSE.LGPL, which must be provided, for details		     **
 ** 									     **
 ** ************************************************************************ **/

#include <strings.h>
#include <stdlib.h>
#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif
#include "strmalloc.h"
#ifdef RKOERROR
#  include "rkoerror.h"
#endif
#ifdef MEMDEBUG
#  include "memdebug.h"
#endif

static char id[3] = "STR";

char *str_nmalloc(char const *in, size_t n) {
	char *out = (char *) NULL;
	size_t sizeit;
#ifdef RKOERROR
	rkoerrno = RKO_OK;
#endif
	if (in) {
		sizeit = n;

		if(!(out = (char *) calloc(sizeit * sizeof(char)
		+ sizeof(size_t) + sizeof(id),1))) {
#ifdef RKOERROR
			(void) rkocpyerror("str_nmalloc : malloc error!");
			rkoerrno = RKOMEMERR;
#endif
		}
		(void) strncpy(out, id, sizeof(id));
		out += sizeof(id);
		(void) memcpy(out, &sizeit, sizeof(size_t));
		out += sizeof(size_t);
		(void) strncpy(out, in, n);
#ifdef RKOERROR
	} else {
			(void) rkocpyerror("str_nmalloc : NULL string!");
			rkoerrno = RKOUSEERR;
#endif
	}
	return out;
}

char *strnmalloc(char const *in, size_t n) {
	return str_nmalloc(in, n);
}

char *str_malloc(char const *in) {
	char *out = (char *) NULL;
	size_t sizeit;

#ifdef RKOERROR
	rkoerrno = RKO_OK;
#endif
	if (!in) {
#ifdef RKOERROR
		(void) rkocpyerror("str_malloc : NULL string!");
		rkoerrno = RKOUSEERR;
#endif
		return out;
	}

	sizeit = strlen(in)+1;
	out = str_nmalloc(in, sizeit);
	if (!out) {	/* test if input is non-zero */
#ifdef RKOERROR
		(void) rkopsterror("str_malloc : ");
#endif
	}
	return out;
}

char *strmalloc(char const *in) {
	return str_malloc(in);
}

/* str_free_ - only use this if sure that a str object is passed
 */
static int str_free_(char **str, size_t len) {
	char *ptr;

	ptr = *str - sizeof(size_t) - sizeof(id);

	(void) memset((void *) ptr, 0, sizeof(id)+sizeof(size_t)+len);
	free((void *) ptr);

	*str = (char *) NULL;

	return 0;
}

int str_free(char **str) {
	size_t sizeit;
#ifdef RKOERROR
	rkoerrno = RKO_OK;
#endif
	if (str && *str) {
		if ((size_t) -1 == (sizeit = str_sizeof(*str))) {
#ifdef RKOERROR
			(void) rkopsterror("str_free : ");
#endif
			return -1;
		}
		return str_free_(str, sizeit);
	}
	return 0;
}

int strfree(char **str) {
	return str_free(str);
}

size_t str_sizeof(const char *str) {
	size_t sizeit;
	char *ptr;

#ifdef RKOERROR
	rkoerrno = RKO_OK;
#endif
	if (str && *str) {
		ptr = (char *) str - sizeof(size_t) - sizeof(id);
		if (strncmp(ptr, id, sizeof(id))) {
#ifdef RKOERROR
			rkoerrno = RKOUSEERR;
		(void) rkocpyerror("str_sizeof : invalid string object!");
#endif
			return (size_t) -1;
		}
		(void) memcpy(&sizeit, (void *) (ptr + sizeof(id)),
			sizeof(size_t));
		return sizeit;
	}
	return 0;
}

/* str_ncpy_ - use this if sure if no size fudge factor is needed */
static char *str_ncpy_(char **str, char *in, size_t n) {
	char *out = (char *) NULL;
	size_t sizeit;

#ifdef RKOERROR
	rkoerrno = RKO_OK;
#endif
	if (!in) return out;

	if ((size_t) -1 == (sizeit = str_sizeof(*str))) {
#ifdef RKOERROR
		(void) rkopsterror("str_ncpy_ : ");
#endif
		return out;
	}

	if (sizeit < n) {	/* insufficient memory */
		str_free_(str, sizeit);
		out = str_nmalloc(in,n);
		return *str = out;
	} else {
		return strncpy(*str, in, n);
	}
}

char *str_ncpy(char **str, char *in, size_t n) {
	return str_ncpy_(str, in, n);
}

char *str_cpy(char **str, char *in) {
	char *retval = (char *) NULL;

	if (!*str || !in)
		return retval;

	retval = str_ncpy_(str, in, strlen(in)+1);
#ifdef RKOERROR
	if (!retval && rkoerrno != RKO_OK)
		rkopsterror("str_cpy : ");
#endif
	return retval;
}

char *str_ncat(char **str, char *in, size_t n) {
	char *out = (char *) NULL;
	size_t sizeit;

#ifdef RKOERROR
	rkoerrno = RKO_OK;
#endif
	if (!in) return out;

	if ((size_t) -1 == (sizeit = str_sizeof(*str))) {
#ifdef RKOERROR
		(void) rkopsterror("str_cat : ");
#endif
		return out;
	}

	if (sizeit <= n + strlen(*str)) {   /* insufficient memory */
		out = str_nmalloc(*str, n + strlen(*str));
		str_free_(str,sizeit);
		strncat(out, in, n);
		return *str = out;
	} else {
		return strcat(*str, in);
	}
}

char *str_cat(char **str, char *in) {
	char *retval = (char *) NULL;

	if (!*str || !in)
		return retval;

	retval = str_ncat(str, in, strlen(in)+1);
#ifdef RKOERROR
	if (!retval && rkoerrno != RKO_OK)
		rkopsterror("str_cat : ");
#endif
	return retval;
}

